skafka/src/main/scala/com/evolutiongaming/skafka/consumer/RebalanceCallback.scala

```diff
@@ -74,101 +74,154 @@ object RebalanceCallback extends RebalanceCallbackInstances with RebalanceCallba
+                  case Failure(exception) =>
+                    Try(fe(exception))
+                  case Success(value) =>
+                    Success(RebalanceCallback.pure(value))
+                }
+
+                loop(Try(source()), s :: ss)
             }
-          case c: Bind[F, _, A1] =>
-            loop(c.source(), c.fs.asInstanceOf[S] :: ss)
-          case Error(a) =>
-            a.raiseError[Try, A1]
         }
       }
 
       for {
-        a <- loop(self, List.empty)
+        a <- loop(Success(self), List.empty)
         a <- Try { a.asInstanceOf[A] }
       } yield a
     }
 
     def toF(
       consumer: RebalanceConsumer
     )(implicit MT: MonadThrowable[F]): F[A] = {
-      type S = Any => RebalanceCallback[F, Any]
+      type S = Try[Any] => Try[RebalanceCallback[F, Any]]
 
-      def continue[A1](c: RebalanceCallback[F, A1], ss: List[S]): F[Any] =
+      def continue[A1](c: Try[RebalanceCallback[F, A1]], ss: List[S]): F[Any] =
         loop(c, ss)
 
       @tailrec
-      def loop[A1](c: RebalanceCallback[F, A1], ss: List[S]): F[Any] = {
+      def loop[A1](c: Try[RebalanceCallback[F, A1]], ss: List[S]): F[Any] = {
         c match {
-          case c: Pure[A1] =>
+          case f @ Failure(a) =>
             ss match {
-              case Nil     => c.a.pure[F].widen
-              case s :: ss => loop(s(c.a), ss)
+              case Nil     => a.raiseError[F, A1].widen
+              case s :: ss => loop(s(f), ss)
             }
-          case c: Lift[F, A1] =>
-            c.fa.widen.flatMap { a =>
-              ss match {
-                case Nil     => a.pure[F].widen
-                case s :: ss => continue(s(a), ss)
-              }
+          case Success(c) =>
+            c match {
+              case c: Pure[A1] =>
+                ss match {
+                  case Nil     => c.a.pure[F].widen
+                  case s :: ss => loop(s(Success(c.a)), ss)
+                }
+              case c: Lift[F, A1] =>
+                c.fa.attempt.flatMap {
+                  case Left(a) =>
+                    ss match {
+                      case Nil     => a.raiseError[F, A1].widen
+                      case s :: ss => continue(s(Failure(a)), ss)
+                    }
+                  case Right(value) =>
+                    ss match {
+                      case Nil     => value.pure[F].widen
+                      case s :: ss => continue(s(Success(value)), ss)
+                    }
+                }
+              case c: WithConsumer[A1] =>
+                // a trick to make `c.f` lazy if F[_] is lazy
+                // by moving it into `flatMap`
+                // using this trick as it does not require Defer/Sync instances
+                val fa = for {
+                  _ <- ().pure[F]
+                  a <- c.f(consumer).fold(_.raiseError[F, A1], _.pure[F])
+                } yield a
+                loop(Success(Lift(fa)), ss)
+              case c: Bind[F, Any, A1] =>
```

Comment:
bind should be moved up in match as this is most common operation

Reply:
done!